src/topology/algebra/filter_basis.lean

```diff
@@ -0,0 +1,400 @@
+  @nhds G (B.topology) (1 : G) = B.to_filter_basis.filter :=
+by { rw B.nhds_eq, simp only [N, one_mul], exact map_id }
+
+@[to_additive]
+lemma nhds_has_basis (B : group_filter_basis G) (x₀ : G) :
+has_basis (@nhds G B.topology x₀) (λ V : set G, V ∈ B) (λ V, (λ y, x₀ * y) '' V)  :=
+by { rw B.nhds_eq, apply B.has_basis }
+
+@[to_additive]
+lemma nhds_one_has_basis (B : group_filter_basis G) :
+has_basis (@nhds G B.topology 1) (λ V : set G, V ∈ B) id  :=
+by { rw B.nhds_one_eq, exact B.to_filter_basis.has_basis }
+
+@[to_additive]
+lemma mem_nhds_one (B : group_filter_basis G) {U : set G} (hU : U ∈ B) : U ∈ @nhds G B.topology 1 :=
+begin
+  rw B.nhds_one_has_basis.mem_iff,
+  exact ⟨U, hU, rfl.subset⟩
+end
+
+/-- If a group is endowed with a topological structure coming from
+a group filter basis then it's a topological group. -/
+@[to_additive, priority 100]
+instance is_topological_group (B : group_filter_basis G) :
+  @topological_group G B.topology _ :=
+begin
+  letI := B.topology,
+  have basis := B.nhds_one_has_basis,
+  have basis' := basis.prod basis,
+  refine topological_group.of_nhds_one _ _ _ _,
+  { rw basis'.tendsto_iff basis,
+    suffices : ∀ U ∈ B, ∃ V W, (V ∈ B ∧ W ∈ B) ∧ ∀ a b, a ∈ V → b ∈ W → a * b ∈ U, by simpa,
+    intros U U_in,
+    rcases mul U_in with ⟨V, V_in, hV⟩,
+    use [V, V, V_in, V_in],
+    intros a b a_in b_in,
+    exact hV ⟨a, b, a_in, b_in, rfl⟩ },
+  { rw basis.tendsto_iff basis,
+    intros U U_in,
+    simpa using inv U_in },
+  { intro x₀,
+    rw [nhds_eq, nhds_one_eq],
+    refl },
+  { intro x₀,
+    rw basis.tendsto_iff basis,
+    intros U U_in,
+    exact conj x₀ U_in }
+end
+
+end group_filter_basis
+
+/-- A `ring_filter_basis` on a ring is a `filter_basis` satisfying some additional axioms.
+  Example : if `R` is a topological ring then the neighbourhoods of the identity are a
+  `ring_filter_basis`. Conversely given a `ring_filter_basis` on a ring `R`, one can define a
+  topology on `R` which is compatible with the ring structure.  -/
+class ring_filter_basis (R : Type u) [ring R] extends add_group_filter_basis R :=
+(mul' : ∀ {U}, U ∈ sets → ∃ V ∈ sets, V * V ⊆ U)
+(mul_left' : ∀ (x₀ : R) {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, x₀*x) ⁻¹' U)
+(mul_right' : ∀ (x₀ : R) {U}, U ∈ sets → ∃ V ∈ sets, V ⊆ (λ x, x*x₀) ⁻¹' U)
+
+namespace ring_filter_basis
+
+variables {R : Type u} [ring R] (B : ring_filter_basis R)
+
+instance : has_mem (set R) (ring_filter_basis R) :=
+⟨λ s B, s ∈ B.sets⟩
+
+lemma mul {U : set R} (hU : U ∈ B) : ∃ V ∈ B, V * V ⊆ U :=
+mul' hU
+
+lemma mul_left (x₀ : R) {U : set R} (hU : U ∈ B) :
+  ∃ V ∈ B, V ⊆ (λ x, x₀*x) ⁻¹' U :=
+mul_left' x₀ hU
+
+lemma mul_right (x₀ : R) {U : set R} (hU : U ∈ B) :
+  ∃ V ∈ B, V ⊆ (λ x, x*x₀) ⁻¹' U :=
+mul_right' x₀ hU
+
+/-- The topology associated to a ring filter basis.
+It has the given basis as a basis of neighborhoods of zero. -/
+def topology : topological_space R := B.to_add_group_filter_basis.topology
```

Comment:
Should this be reducible per `[reducible non-instances]`?

Reply:
If I understand this note correctly, this reducibility attribute should be added in cases where not adding it causes type class inference failure. So I would say it's not needed here (this definition is used successfully in upcoming PRs without this attribute).

Reply:
@fpvandoorn, thoughts?